forc-plugins/forc-client/src/util/pkg.rs

```diff
@@ -1,8 +1,142 @@
 use anyhow::Result;
 use forc_pkg::manifest::GenericManifestFile;
 use forc_pkg::{self as pkg, manifest::ManifestFile, BuildOpts, BuildPlan};
-use pkg::{build_with_options, BuiltPackage};
+use forc_util::user_forc_directory;
+use pkg::{build_with_options, BuiltPackage, PackageManifestFile};
+use std::fs::File;
+use std::io::{Read, Write};
+use std::path::PathBuf;
 use std::{collections::HashMap, path::Path, sync::Arc};
+use sway_utils::{MAIN_ENTRY, MANIFEST_FILE_NAME, SRC_DIR};
+
+/// The name of the folder that forc generated proxy contract project will reside at.
+pub const PROXY_CONTRACT_FOLDER_NAME: &str = ".generated_proxy_contracts";
+/// Forc.toml for the default proxy contract that 'generate_proxy_contract_src()' returns.
+pub const PROXY_CONTRACT_FORC_TOML: &str = r#"
+[project]
+authors = ["Fuel Labs <contact@fuel.sh>"]
+entry = "main.sw"
+license = "Apache-2.0"
+name = "proxy_contract"
+
+[dependencies]
+standards = { git = "https://github.com/FuelLabs/sway-standards/" }
+"#;
+
+/// Generates source code for proxy contract owner set to the given 'addr'.
+pub(crate) fn generate_proxy_contract_src(addr: &str, impl_contract_id: &str) -> String {
+    format!(
+        r#"
+contract;
+
+use std::execution::run_external;
+use standards::src5::{{AccessError, SRC5, State}};
+use standards::src14::SRC14;
+
+/// The owner of this contract at deployment.
+const INITIAL_OWNER: Identity = Identity::Address(Address::from({addr}));
+
+// use sha256("storage_SRC14") as base to avoid collisions
+#[namespace(SRC14)]
+storage {{
+    // target is at sha256("storage_SRC14_0")
+    target: ContractId = ContractId::from({impl_contract_id}),
+    owner: State = State::Initialized(INITIAL_OWNER),
+}}
+
+impl SRC5 for Contract {{
+    #[storage(read)]
+    fn owner() -> State {{
+        storage.owner.read()
+    }}
+}}
+
+impl SRC14 for Contract {{
+    #[storage(write)]
+    fn set_proxy_target(new_target: ContractId) {{
+        only_owner();
+        storage.target.write(new_target);
+    }}
+}}
+
+#[fallback]
+#[storage(read)]
+fn fallback() {{
+    // pass through any other method call to the target
+    run_external(storage.target.read())
+}}
+
+#[storage(read)]
+fn only_owner() {{
+    require(
+        storage
+            .owner
+            .read() == State::Initialized(msg_sender().unwrap()),
+        AccessError::NotOwner,
+    );
+}}
+"#
+    )
+}
+
+/// Updates the given package manifest file such that the address field under the proxy table updated to the given value.
+/// Updated manifest file is written back to the same location, without thouching anything else such as comments etc.
+/// A safety check is done to ensure the proxy table exists before attempting to udpdate the value.
+pub(crate) fn update_proxy_address_in_manifest(
+    address: &str,
+    manifest: &PackageManifestFile,
+) -> Result<()> {
+    let mut toml = String::new();
```

Comment:
It might be cleaner to add this logic in a a `write_to_file()` in `PackageManifest`'s impl, as  there is already a `from_file()` there.

Then this function could be simpler: 

```rust
    pub(crate) fn update_proxy_address_in_manifest(
        address: &str,
        manifest: &PackageManifestFile,
    ) -> Result<()> {
           match manifest.proxy() {
               Ok(_) => {
                   let mut new_manifest = manifest.clone();
                   new_manifest.proxy.unwrap().address = address;
                   new_manifest.write_to_file();
                   Ok(())
               }
               Err(_) => // return error?
           }
}
```

Reply:
https://github.com/FuelLabs/sway/issues/6074 -- This PR is getting very big, i will do it on a clean-up run after this one https://github.com/FuelLabs/sway/issues/6074